bird-house/birdhouse-deploy
# Make all components pluggable

## Overview

Makes all components part of the "pluggable architecture" meaning that no components are required to be configured, each components can be added to the stack as needed. 

See previous discussion for this PR at #295 

## Changes

**Non-breaking changes**

The components that will be added to the stack are only those whose configuration directory is listed in either `DEFAULT_CONF_DIRS` or `EXTRA_CONF_DIRS`.

**Breaking changes**

None

## Related Issue / Discussion

- Resolves #213 

## Additional Information

In a separate PR we should:

- components configurations are now stored in `configs`, `components`, and `optional_components` directories. Now that these distinctions are no longer relevant it might make more sense to divide them as follows:
    - `components`: all components (optional or otherwise; eg: finch, weaver, twitcher)
    - `shared_dependencies`: docker containers required to run multiple components (eg: postgres, mongodb)
    - `component_enhancements`: additional configurations that modify an existing component (eg: all-public-access, database-external-ports)

## Tests
- Ensure baseline status on `master` branch pass Jenkins to prepare for upgrade test
  - If env was previous updated to newer branch, run this script to delete new ignored files that branch `master` do not know about, which would break autodeploy [delete-ignore.sh.txt](https://github.com/bird-house/birdhouse-deploy/files/11036499/delete-ignore.sh.txt)
  - First `./pavics-compose.sh up -d` [master-up.txt](https://github.com/bird-house/birdhouse-deploy/files/11036376/master-up.txt)
  - First `./pavics-compose.sh ps` not all container healthy [master-ps.txt](https://github.com/bird-house/birdhouse-deploy/files/11036378/master-ps.txt)
  - Subsequent `./pavics-compose.sh up -d` for all services responding so Weaver can add provider [master2-up.txt](https://github.com/bird-house/birdhouse-deploy/files/11036380/master2-up.txt)
  - Subsequent `./pavics-compose.sh ps` ready to run Jenkins [master-ps-2.txt](https://github.com/bird-house/birdhouse-deploy/files/11036479/master-ps-2.txt)
  - Jenkins output, only known expected failure with `esgf-dap.ipynb` and `WMS_example.ipynb` [baseline-master-PAVICS-e2e-workflow-tests-current-production-version-114-consoleText.txt](https://github.com/bird-house/birdhouse-deploy/files/11036409/baseline-master-PAVICS-e2e-workflow-tests-current-production-version-114-consoleText.txt)
- Upgrade test via autodeploy to simulate production environment rollout
  - Autodeploy logs which includes `./pavics-compose.sh up -d` [autodeploy.log](https://github.com/bird-house/birdhouse-deploy/files/11036433/autodeploy.log)
  - `./pavics-compose.sh ps` after autodeploy [new-ps-after-autodeploy.txt](https://github.com/bird-house/birdhouse-deploy/files/11036463/new-ps-after-autodeploy.txt)
- Ensure no regressions, no backward incompatible changes, when using default config after upgrade
  - Jenkins output, same known expected failure with `esgf-dap.ipynb` and `WMS_example.ipynb` [new-branch-PAVICS-e2e-workflow-tests-current-production-version-115-consoleText.txt](https://github.com/bird-house/birdhouse-deploy/files/11036466/new-branch-PAVICS-e2e-workflow-tests-current-production-version-115-consoleText.txt)
  - Diff `master-ps-2.txt` and `new-ps-after-autodeploy.txt`, same number of containers so nothing missing (default enabled components are good)
  - Trigger autodeploy again, it works, no missing `.gitignore`
- (not done yet) Test the new ability provided by this PR to pick and choose a subset of components
  - Proxy, Finch, Weaver only: the subset that CRIM runs but CRIM unable to use PAVICS because it does not want to rest of PAVICS stack
  - Proxy, Thredds, Twitcher, Magpie, Phoenix only + external birds: the subset that PCIC needs but is stuck with the entire PAVICS stack
